### PR TITLE
Support using multiple GPUs to do mining together

### PR DESCRIPTION
**Pull request by ShengguangXiao**
## Description

Support using multiple GPUs to do mining together

Update the POW library. Create a new thread for each GPU to do mining.

https://github.com/Zilliqa/Zilliqa/issues/555

## Review Suggestion

Check the libPOW.

Use a PC with more than 1 GPUs, set NUM_DEVICE_TO_USE in constants to 2. If you are using OpenCL GPU, set OPENCL_GPU_MINE to true, if you are using CUDA gpu, set CUDA_GPU_MINE to true, build the software with ./build.sh opencl or ./build.sh cuda command. Then run cd build && ./tests/POW/Test_POW see if the test case can pass.

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [x] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

**Comment by ShengguangXiao** [comment]
Run 20 nodes test, 2 nodes using 2 AMD gpu together to do PoW, run to 23 DS Blocks and 105 Tx Blocks without problem.

File: src/libPOW/pow.cpp [hunk 1]
```diff
@@ -250,40 +251,89 @@ ethash_mining_result_t POW::MineFull(ethash_full_t& full,
...
     dev::eth::WorkPackage wp;
     wp.blockNumber = blockNum;
     wp.boundary = (dev::h256)(dev::u256)((dev::bigint(1) << 256)
                                          / (dev::u256(1) << difficulty));
 
     wp.header = dev::h256{header_hash.b, dev::h256::ConstructFromPointer};
-    wp.startNonce = std::time(0);
+
+    constexpr uint32_t NONCE_SEGMENT_WIDTH = 40;
+    const uint64_t NONCE_SGEMENT = (uint64_t)pow(2, NONCE_SEGMENT_WIDTH);
```
Comment (Gnnng) [inline src/libPOW/pow.cpp]: typo `NONCE_SGEMENT` --> `NONCE_SEGMENT`

File: src/libPOW/pow.cpp [hunk 2]
```diff
@@ -250,40 +251,89 @@ ethash_mining_result_t POW::MineFull(ethash_full_t& full,
...
+                                << index);
     dev::eth::WorkPackage wp;
     wp.blockNumber = blockNum;
     wp.boundary = (dev::h256)(dev::u256)((dev::bigint(1) << 256)
                                          / (dev::u256(1) << difficulty));
 
     wp.header = dev::h256{header_hash.b, dev::h256::ConstructFromPointer};
-    wp.startNonce = std::time(0);
+
+    constexpr uint32_t NONCE_SEGMENT_WIDTH = 40;
```
Comment (Gnnng) [inline src/libPOW/pow.cpp]: perhaps `const` will be good enough here. Just for code tidiness.

File: src/libPOW/pow.h [hunk 3]
```diff
@@ -102,7 +102,11 @@ class POW
     ethash_light_t ethash_light_client;
     uint64_t currentBlockNum;
     std::atomic<bool> m_shouldMine;
-    std::unique_ptr<dev::eth::Miner> m_miner;
+    std::vector<dev::eth::MinerPtr> m_miners;
+    std::vector<ethash_mining_result_t> m_vecMiningResult;
+    std::atomic<int> m_minerIndex;
+    std::condition_variable m_cvMineResult;
```
Comment (Gnnng) [inline src/libPOW/pow.h]: perhaps we can be consistent on the namings? Using the word `MiningResult` is better.

File: src/libPOW/pow.h [hunk 4]
```diff
@@ -102,7 +102,11 @@ class POW
...
+    std::mutex m_mutexMineResult;
```
Comment (Gnnng) [inline src/libPOW/pow.h]: The same naming suggestion here, `MiningResult`

File: src/libPOW/pow.h (hunk 3 above)
Comment (ShengguangXiao) [reply src/libPOW/pow.h]: Ok, I will update it.

**Pull request merged**